# MLTree Pipeline Description

This workflow has a few steps, which can be divided into a 1) multiple sequence alignment (MSA) workflow, 2) ModelTest workflow to determine the best nucleotide or amino acid substitution model, 3) the BuildTree workflow for maximum likelihood (ML) tree construction, and 4) the RepSelection workflow to pick out representative sequences for the bayesian tree using (Parnas)[https://github.com/flu-crew/parnas] 

Two separate ML trees were made, an all-Picornaviridae nucleotide tree and an only-kobuvirus nucleotide tree:

- I downloaded Picornaviridae reference sequences within NCBI Virus within a minimum sequence length of 7200bp and 9000bp with complete nucleotide completeness (Last Access: 6/23/2022) (taxid: 12058; 130 hits). These sequences, along with our full-length Malagasy kobuvirus sequence were aligned and imported into Geneious to exclude gaps at either alignment end which could interfere with tree analyses. The script used to rename FASTA headers can be found in the scripts directory within 1_MSA, which was necessary before uploading the MSA to RAxML as RAxML won't accept erroneous characters in sequence names. I additionally searched for bat picornaviruses that may have been missed by this broader Picornaviridae search (Bat Picornavirus, taxid 1281456, Bat Picornavirus 1, taxid 1074863, Bat Picornavirus 2, taxid 1074864, Bat Picornavirus 3, taxid 1074865, Washington bat picornavirus, taxid 1888309, Aichivirus F, taxid 1986959, Aichivirus F1, taxid 2758106, Aichivirus F2, 2755032, Miniopterus picornavirus 1, taxid 2184394, Miniopterus picornavirus 2, taxid 2184395)
  
- I downloaded kobuvirus  sequences within NCBI Virus with a minimum sequence length of 5200bp and a max sequence length of 9000bp and complete nucleotide completeness (Last Access: 6/23/2022) (taxid: 194960; 109 hits). These sequences, along with our full-length and partial-length Malagasy kobuvirus sequences were aligned and imported into Geneious to exclude gaps at either alignment end which could interfere with tree analyses. The script used to rename FASTA headers can be found in the scripts directory within 1_MSA, which was necessary before uploading the MSA to RAxML as RAxML won't accept erroneous characters in sequence names
